hspell: fix cross-platform build issues

kampka opened the pull request:
###### Motivation for this change

hspell uses a couple of binaries it generates during it's own build.
These will cause the build to fail later in cross-platform builds as
these binaries have the wrong executable format.
Using a hspell to bootstrap it's own build natively resolves this issue.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

kampka commented:
@GrahamcOfBorg build hspell
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.hspell

kevincox commented:
I notice that this doesn't have a maintainer listed. Would you be interested in taking it over?

kampka commented:
> I notice that this doesn't have a maintainer listed. Would you be interested in taking it over?

Sure, added myself as a maintainer, but in all fairness, I'm not passionate about this at all.
I mainly fixed the issue because the package is in the build chain of something else I am trying to cross-build.
Since this package seems low-maintenance without new releases in quite a while, I can probably keep an eye on it though it won't become a priority.

kampka commented:
@GrahamcOfBorg build hspell
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.hspell

jonringer commented:
@GrahamcOfBorg eval

matthewbauer commented on pkgs/development/libraries/hspell/default.nix:
```diff
@@ -1,27 +1,62 @@
   };
-}
+
+  bootstrap = stdenv.mkDerivation (mkDrv{ pname = "hspell-bootstrap"; } // {
```
This shouldn't need to be in a separate derivation. I think you can just do something like:

```
preBuild = ''
  make CC=${buildPackages.stdenv.cc}/bin/cc find_sizes
'';
```

to avoid it. Ideally we should patch hspell to do the right thing (use $(CC) for hspell and $(BUILD_CC) for find_sizes).

matthewbauer replied:
But building a bootstrap like this is not idiomatic, since lots of other packages have a find_sizes that they get correctly.

kampka replied:
> Ideally we should patch hspell to do the right thing (use $(CC) for hspell and $(BUILD_CC) for find_sizes).

Agreed this should be fixed upstream, but as it does not seem to be an actively maintained project, I'd rather have a working version here in any case.

> This shouldn't need to be in a separate derivation.

Done.

kampka committed:
hspell: fix cross-platform build issues

hspell uses a couple of binaries it generates during it's own build.
These will cause the build to fail later in cross-platform builds as
these binaries have the wrong executable format.
Using a hspell to bootstrap it's own build natively resolves this issue.

kampka commented:
@GrahamcOfBorg build hspell
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.hspell

Ericson2314 commented:
Is there any way to contribute patches upstream or is this project pure cathedral style?

kampka commented:
> Is there any way to contribute patches upstream or is this project pure cathedral style?

I am neither familiar with the term `cathedral style` nor the project, really.
Could you clarify what you mean?

alyssais commented:
> I am neither familiar with the term `cathedral style` nor the project, really.

Cathedral style means software that is open source, but largely or entirely directed by one person or group. It’s less about whether they accept outside patches, but more about whether outsiders can be part of changing how the program works or coming up with its overall direction. The term comes from a famous essay about open source software development called <cite>The Cathedral and the Bazaar</cite>.

kampka commented:
> Cathedral style means software that is open source, but largely or entirely directed by one person or group. It’s less about whether they accept outside patches, but more about whether outsiders can be part of changing how the program works or coming up with its overall direction. The term comes from a famous essay about open source software development called The Cathedral and the Bazaar.

Got it, thanks for clarifying.

jtojnar commented:
I tried updating the package to the latest version:

https://github.com/jtojnar/nixpkgs/pull/new/hspell-1.4

Could you please try if the situation is not better there?

kampka commented:
> I tried updating the package to the latest version:
> 
> https://github.com/jtojnar/nixpkgs/pull/new/hspell-1.4
> 
> Could you please try if the situation is not better there?

I tried that already, but it not only has the same issues, it actually introduces new ones when linking against `zlib` that I don't currently have the time to figure out.

symphorien commented:
@GrahamcOfBorg build hspell
@GrahamcOfBorg build pkgsCross.aarch64-multiplatform.hspell
